js/src/analysis/data.js

```diff
@@ -0,0 +1,104 @@
+/* global wp YoastSEO */
+import debounce from "lodash/debounce";
+
+let data = {};
+let isDirty;
+
+/**
+ * Checks whether the current data and the Gutenberg data are the same.
+ *
+ * @param {Object} currentData The current data.
+ * @param {Object} gutenbergData The data from Gutenberg.
+ * @returns {boolean} Whether the current data and the gutenbergData is the same.
+ */
+const isShallowEqual = ( currentData, gutenbergData ) => {
+	for( let dataPoint in currentData ) {
+		if ( currentData.hasOwnProperty( dataPoint ) ) {
+			if( ! ( dataPoint in gutenbergData ) || currentData[ dataPoint ] !== gutenbergData[ dataPoint ] ) {
+				return false;
+			}
+		}
+	}
+	for( let dataPoint in gutenbergData ) {
+		if ( gutenbergData.hasOwnProperty( dataPoint ) ) {
+			if( ! ( dataPoint in currentData ) || currentData[ dataPoint ] !== gutenbergData[ dataPoint ] ) {
+				return false;
+			}
+		}
+	}
+	return true;
+};
+
+/**
+ * Updates the data object.
+ *
+ * @param {Object} data The current data.
+ * @param {Object} gutenbergData The data from Gutenberg.
+ * @returns {Object} The data.
+ */
+const updateData = ( data, gutenbergData ) => {
+	// Set isDirty to false if the current data and Gutenberg data are unequal.
+	isDirty = ! isShallowEqual( data, gutenbergData );
+
+	// If there is new data from Gutenberg, overwrite the data object with the new data.
+	if ( isDirty ) {
+		data = gutenbergData;
+	}
+	return data;
+};
+
+/**
+ * Collects the content, title, slug and excerpt of a post from Gutenberg.
+ *
+ * @returns {{content: string, title: string, slug: string, excerpt: string}} The content, title, slug and excerpt.
+ */
+const collectGutenbergData =  () => {
+	return {
+		content: wp.data.select( "core/editor" ).getEditedPostAttribute( "content" ),
+		title: wp.data.select( "core/editor" ).getEditedPostAttribute( "title" ),
+		slug: wp.data.select( "core/editor" ).getEditedPostAttribute( "slug" ),
+		excerpt: wp.data.select( "core/editor" ).getEditedPostAttribute( "excerpt" ),
+	};
+};
+
+/**
+ * Refreshes YoastSEO's app when the Gutenberg data is dirty.
+ *
+ * @returns {void}
+ */
+const refreshYoastSEO = () => {
+	let gutenbergData = collectGutenbergData();
+	data = updateData( data, gutenbergData );
```

Comment:
It seems to me that you could do the isDirty check here and then you could overwrite data only if needed. That way you do not need a separate updateDate function.